iroh-router/src/router.rs

```diff
@@ -0,0 +1,211 @@
+        Self {
+            endpoint,
+            protocols: ProtocolMap::default(),
+        }
+    }
+
+    pub fn accept(mut self, alpn: Vec<u8>, handler: Arc<dyn ProtocolHandler>) -> Self {
+        self.protocols.insert(alpn, handler);
+        self
+    }
+
+    /// Returns the [`Endpoint`] of the node.
+    pub fn endpoint(&self) -> &Endpoint {
+        &self.endpoint
+    }
+
+    /// Returns a protocol handler for an ALPN.
+    ///
+    /// This downcasts to the concrete type and returns `None` if the handler registered for `alpn`
+    /// does not match the passed type.
+    pub fn get_protocol<P: ProtocolHandler>(&self, alpn: &[u8]) -> Option<Arc<P>> {
+        self.protocols.get_typed(alpn)
+    }
+
+    pub async fn spawn(self) -> Result<Router> {
+        // Update the endpoint with our alpns.
+        let alpns = self
+            .protocols
+            .alpns()
+            .map(|alpn| alpn.to_vec())
+            .collect::<Vec<_>>();
+
+        let protocols = Arc::new(self.protocols);
+        if let Err(err) = self.endpoint.set_alpns(alpns) {
+            shutdown(&self.endpoint, protocols.clone()).await;
+            return Err(err);
+        }
+
+        let mut join_set = JoinSet::new();
+        let endpoint = self.endpoint.clone();
+        let protos = protocols.clone();
+        let cancel = CancellationToken::new();
+        let cancel_token = cancel.clone();
+
+        let fut = async move {
+            let protocols = protos;
+            loop {
+                tokio::select! {
+                    biased;
+                    _ = cancel_token.cancelled() => {
+                        break;
+                    },
+                    // handle incoming p2p connections.
+                    Some(incoming) = endpoint.accept() => {
+                        let protocols = protocols.clone();
+                        join_set.spawn(async move {
+                            handle_connection(incoming, protocols).await;
+                            anyhow::Ok(())
+                        });
+                    },
+                    // handle task terminations and quit on panics.
+                    res = join_set.join_next(), if !join_set.is_empty() => {
+                        match res {
+                            Some(Err(outer)) => {
+                                if outer.is_panic() {
+                                    error!("Task panicked: {outer:?}");
+                                    break;
+                                } else if outer.is_cancelled() {
+                                    debug!("Task cancelled: {outer:?}");
+                                } else {
+                                    error!("Task failed: {outer:?}");
+                                    break;
+                                }
+                            }
+                            Some(Ok(Err(inner))) => {
+                                debug!("Task errored: {inner:?}");
```

Comment:
This is an `anyhow::Error` I think, should use alt formatting for those when logging: `{inner:#}`.

Possibly the others are also better at alt formatting?  But I'm not sure, would have to try it.  For the `Task panicked` we certainly want to make sure to log the backtrace.